### PR TITLE
Rereapply optional ScalarType interface changes that were reverted in D16079809

### PR DESCRIPTION
**Pull request by nairbv**
re-apply changes reverted in:
https://github.com/pytorch/pytorch/pull/22412

Also change log_softmax to take positional arguments. Long-term we do want the kwarg-only interface, but seems to currently be incompatible with jit serialization.

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Can you add a test for `log_softmax` to test this specific positional argument/ kw-only argument problem? I would like to ensure we notice this breakage next time if we change the API again

**Comment by nairbv** [comment]
FYI @ailzhang, when I re-land this it'll re-break XLA

**Comment by nairbv** [comment]
> Can you add a test for `log_softmax` to test this specific positional argument/ kw-only argument problem? I would like to ensure we notice this breakage next time if we change the API again

I think the test we'll want will be in jit? 

@gchanan

**Comment by wanchaol** [comment]
> > Can you add a test for `log_softmax` to test this specific positional argument/ kw-only argument problem? I would like to ensure we notice this breakage next time if we change the API again
> 
> I think the test we'll want will be in jit?
> 
> @gchanan

While I think we should add to the somewhere that JIT tests and Autograd tests can share with each other, like `common_method_invocations.py`, we will want to ensure the eager API change be captured as well.

**Comment by ailzhang** [comment]
@nairbv we should add `dtype` tests for all affected apis in common_method_invocations to prevent future breakage, e.g `prod/cumsum/...` etc (I think currently we don't have dtype tests coverage for them.)

**Comment by nairbv** [comment]
> @nairbv we should add `dtype` tests for all affected apis in common_method_invocations to prevent future breakage, e.g `prod/cumsum/...` etc (I think currently we don't have dtype tests coverage for them.)

I'm happy to add tests, but it seems like if someone is explicitly changing the API (as I was) they might just also change the tests. If we had a test that attempted to load a model and failed, it'd be much clearer that something was meaningfully incorrect. Asserting that an API is what the API says it is seems a bit redundant.

**Comment by ailzhang** [comment]
@nairbv  Yea I agree. We really should have had test coverage for those so that we knew it's BC breaking. I'm only requesting adding tests here to prevent future breakage :D

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1192,11 +1180,7 @@
     CPU: logspace_cpu_out
     CUDA: logspace_cuda_out
 
-# FIXME: These could be combined as optional<ScalarType> but for https://github.com/pytorch/pytorch/issues/6593.
-- func: log_softmax(Tensor self, int dim, ScalarType dtype) -> Tensor
-  variants: function, method
-
-- func: log_softmax(Tensor self, int dim) -> Tensor
+- func: log_softmax(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: we should comment why this and softmax don't follow the usual convention.
